src/software/logger/proto_logger.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include <google/protobuf/message.h>
+
+#include <functional>
+#include <string>
+#include <thread>
+
+#include "software/multithreading/thread_safe_buffer.hpp"
+
+/**
+ * Logs incoming Protobufs to a folder to be played back later.
+ *
+ * Each entry will contain:
+ *  - The timestamp
+ *  - The protobuf type
+ *  - The protobuf serialized as base64 (to remove newline characters)
+ *
+ * Stored in log_path/proto_YYYY_MM_DD_HH_MM_SS/
+ * With each entry in the file formatted as timestamp,protobuf_type,protobuf_base64
+ * Note that in order to reduce the size of the log files, the files are compressed
+ * using gzip.
+ *
+ * We need to store the data in a way that we can:
+ *  1. Replay the data chronologically
+ *  2. Seek to a specific time (random access)
+ *
+ * To seek to a specific time, we need to load the entire log file into memory.
+ * To make this feasible, we store the data in chunks. Each chunk contains
+ * REPLAY_MAX_CHUNK_SIZE_BYTES of serialized protos.
+ * We can load each chunk into memory and perform search operations to find the
+ * appropriate entry. Or we can just play the chunks in order.
+ */
+class ProtoLogger
+{
+    /**
+     * A struct to store a serialized protobuf message for the purpose of replay logging
+     */
+    struct SerializedProtoLog
+    {
+        std::string protobuf_type_full_name;
+        std::string serialized_proto;
+        double receive_time_sec;
+    };
+
+   public:
+    /**
+     * Constructor
+     * @param log_path The path to the directory where the logs will be saved
+     * @param time_provider A function that returns the current time in seconds
+     * @param friendly_colour_yellow Whether the friendly team is yellow or not
+     */
+    explicit ProtoLogger(const std::string& log_path,
+                         std::function<double()> time_provider,
+                         bool friendly_colour_yellow);
+
```

Comment:
I think it is a good idea to delete the copy constructor 

Something like the following: 

```
ProtoLogger(const ProtoLogger&)            = delete;
ProtoLogger& operator=(const ProtoLogger&) = delete;
```

We don't really want to copy this class since it is heap allocated. If you do want to copy this class, I think you should create your own copy constructor because of threading issues.

Reply:
Removed